File: backend/src/atsresume/pipeline/steps.py
"""Pipeline steps, in the order they run."""

from __future__ import annotations

import logging
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field

from .. import store
from ..config import get_settings
from ..llm import structured
from ..models import (
    AtsReport,
    GapAnalysis,
    JobSpec,
    ResumeFacts,
    Strategy,
    TailoredResume,
    TruthReport,
)
from ..truth.entailment import check_entailment
from ..truth.guard import entailment_pairs, merge_violations, run_truth_guard
from . import prompts
from .sanitize import sanitize
from .score import compute_ats_report

logger = logging.getLogger(__name__)


def _block(tag: str, body: str) -> str:
    return f"<{tag}>\n{body}\n</{tag}>"


def extract_resume_facts(raw_resume_text: str) -> ResumeFacts:
    cached = store.get_facts(raw_resume_text)
    if cached is not None:
        return cached

    facts = structured(
        system=prompts.RESUME_EXTRACTION,
        user="Extract structured facts from this resume.\n\n"
        + _block("resume", raw_resume_text),
        schema=ResumeFacts,
        step="extract",
    )
    store.put_facts(raw_resume_text, facts)
    return facts


def extract_job_spec(jd_text: str, source_note: str) -> JobSpec:
    return structured(
        system=prompts.JD_EXTRACTION,
        user=f"Decompose this job description.\n\nSource: {source_note}\n\n"
        + _block("job_description", jd_text),
        schema=JobSpec,
        step="analyze",
    )


def analyze_gaps(job: JobSpec, facts: ResumeFacts) -> GapAnalysis:
    return structured(
        system=prompts.GAP_ANALYSIS,
        user="Compare this candidate against this role.\n\n"
        + _block("job_spec", job.model_dump_json(indent=2))
        + "\n\n"
        + _block("resume_facts", facts.model_dump_json(indent=2)),
        schema=GapAnalysis,
        step="gaps",
    )


@dataclass
class TailorOutcome:
    tailored: TailoredResume
    truth: TruthReport
    repair_attempted: bool
    # Why the first draft was rejected, if it was. The repair round doubles the
    # cost of the most expensive step, so knowing which rule keeps tripping is
    # the difference between tuning the prompt and guessing at it.
    first_draft_violations: list[str] = field(default_factory=list)


def tailor_resume(
    job: JobSpec,
    facts: ResumeFacts,
    gaps: GapAnalysis,
    raw_resume_text: str,
) -> TailorOutcome:
    """Rewrite, then verify. A failing draft gets exactly one repair attempt.

    One, not many: a loop that keeps pushing until the guard passes is an
    optimiser applying pressure toward whatever wording slips past it, which is
    the opposite of what the guard is for. If a second draft still fails, the
    candidate is shown precisely which lines are unverifiable rather than handed
    a resume that reads well and cannot be defended in an interview.
    """
    jd_terms = job.all_terms()

    brief = "\n\n".join(
        [
            "Rewrite this candidate's resume for this role.",
            _block("job_spec", job.model_dump_json(indent=2)),
            _block("resume_facts", facts.model_dump_json(indent=2)),
            _block("gap_analysis", gaps.model_dump_json(indent=2)),
            "The gap analysis lists missing_keywords the candidate cannot truthfully claim. "
            "Do not use them.\n"
            "Surface every recoverable_keyword - those are already true and merely buried.",
        ]
    )

    def verify(draft: TailoredResume) -> TruthReport:
        """Token checks first, then entailment on what survives them.

        Ordering is a cost decision as much as a correctness one. The token
        checks are free and catch the loud fabrications; running the model over
        a draft that already has an invented metric in it would be paying to
        learn something we already know.
        """
        report = run_truth_guard(draft, facts, raw_resume_text, jd_terms)
        if not get_settings().enable_entailment:
            return report
        return merge_violations(report, check_entailment(entailment_pairs(draft, facts)))

    tailored = sanitize(
        structured(
            system=prompts.TAILOR,
            user=brief,
            schema=TailoredResume,
            step="tailor",
        )
    )
    truth = verify(tailored)
    repair_attempted = False
    first_draft_violations: list[str] = []

    if not truth.passed:
        repair_attempted = True
        # Log the detail, not the line. The detail names the term that tripped
        # the rule, which is the only part that says whether this was a real
        # catch or the guard being too eager.
        first_draft_violations = [
            f"{v.code.value}: {v.detail}" for v in truth.violations if v.severity == "error"
        ]
        findings = "\n".join(
            f'- [{v.code.value}] {v.location}: {v.detail}\n  Line: "{v.offending}"'
            for v in truth.violations
            if v.severity == "error"
        )
        logger.warning(
            "Truth guard rejected the first draft (%d errors): %s",
            truth.error_count,
            "; ".join(first_draft_violations[:5]),
        )

        repair_brief = "\n\n".join(
            [
                brief,
                "Your previous draft failed verification against the uploaded resume:\n" + findings,
                "Rewrite it. For each failing line, either restate it using only what its "
                "sources actually say, or drop it. Do not try to justify a figure or a "
                "technology that is not in the original resume.",
            ]
        )
        tailored = sanitize(
            structured(
                system=prompts.TAILOR,
                user=repair_brief,
                schema=TailoredResume,
                step="tailor",
            )
        )
        truth = verify(tailored)

    return TailorOutcome(
        tailored=tailored,
        truth=truth,
        repair_attempted=repair_attempted,
        first_draft_violations=first_draft_violations,
    )


def strategize(
    job: JobSpec,
    gaps: GapAnalysis,
    tailored: TailoredResume,
    report: AtsReport,
) -> Strategy:
    return structured(
        system=prompts.STRATEGY,
        user="Advise this candidate on this application.\n\n"
        + _block("job_spec", job.model_dump_json(indent=2))
        + "\n\n"
        + _block("gap_analysis", gaps.model_dump_json(indent=2))
        + "\n\n"
        + _block("tailored_resume", tailored.model_dump_json(indent=2))
        + "\n\n"
        + _block("computed_ats_score", report.model_dump_json(indent=2)),
        schema=Strategy,
        step="strategy",
    )


@dataclass
class PipelineResult:
    job: JobSpec
    facts: ResumeFacts
    gaps: GapAnalysis
    tailored: TailoredResume
    truth: TruthReport
    report: AtsReport
    strategy: Strategy
    repair_attempted: bool


def run_pipeline(raw_resume_text: str, jd_text: str, source_note: str) -> PipelineResult:
    """Full run. The two extractions are independent, so they overlap.

    The SDK client is thread-safe and both calls are I/O bound, so a two-worker
    pool halves the slowest serial segment for the cost of one import.
    """
    with ThreadPoolExecutor(max_workers=2) as pool:
        job_future = pool.submit(extract_job_spec, jd_text, source_note)
        facts_future = pool.submit(extract_resume_facts, raw_resume_text)
        job = job_future.result()
        facts = facts_future.result()

    gaps = analyze_gaps(job, facts)
    outcome = tailor_resume(job, facts, gaps, raw_resume_text)
    report = compute_ats_report(job, facts, outcome.tailored)
    strategy = strategize(job, gaps, outcome.tailored, report)

    return PipelineResult(
        job=job,
        facts=facts,
        gaps=gaps,
        tailored=outcome.tailored,
        truth=outcome.truth,
        report=report,
        strategy=strategy,
        repair_attempted=outcome.repair_attempted,
    )

File: apps/web/src/lib/pipeline/steps.ts
/**
 * The pipeline steps, in the order they run.
 *
 * Port of backend/src/atsresume/pipeline/steps.py, with three deliberate
 * changes, each one a finding from reviewing the original:
 *
 * 1. **Strategy is off the critical path.** It is advice about a resume, not
 *    part of one, and it was costing roughly twenty seconds of a candidate
 *    staring at a progress bar before they could see anything. It is now a
 *    separate call the editor makes once it is already open.
 *
 * 2. **Every step is cached by content.** One resume against ten postings
 *    extracts once. See cache.ts for the privacy rule that decides which
 *    steps may share a cache entry across users and which may not.
 *
 * 3. **Verification is a service call, not an import.** The guard is hundreds
 *    of lines of rules over a technology vocabulary, and owning two copies of
 *    it in two languages would mean owning two sets of rules that disagree
 *    under pressure. It stays in Python; this calls it.
 */

import {
  GapAnalysisSchema,
  JobSpecSchema,
  ResumeFactsSchema,
  TailoredResumeSchema,
  computeAtsReport,
  sanitize,
  type AtsReport,
  type GapAnalysis,
  type JobSpec,
  type ResumeFacts,
  type TailoredResume,
  type TruthReport,
} from "@ats/core";

import { configFor } from "../llm/profiles";
import { structured, type Usage } from "../llm/structured";
import { cacheKey, withCache, type CacheScope } from "./cache";
import { PROMPTS, block } from "./prompts";
import { runGuard } from "./guard-client";

export type StepTrace = {
  step: string;
  model: string;
  ms: number;
  usage: Usage | null;
  cacheHit: boolean;
};

export type Traced<T> = { value: T; trace: StepTrace };

async function runStep<T>(opts: {
  step: Parameters<typeof configFor>[0];
  promptVersion: string;
  system: string;
  user: string;
  schema: Parameters<typeof structured>[0]["schema"];
  schemaName: string;
  scope: CacheScope;
  signal?: AbortSignal;
}): Promise<Traced<T>> {
  const config = configFor(opts.step);
  const started = Date.now();

  const key = cacheKey({
    step: opts.step,
    promptVersion: opts.promptVersion,
    model: config.model,
    // The system prompt is in the version, so only the variable half needs
    // hashing. Hashing both would invalidate on a whitespace edit that the
    // version deliberately did not bump for.
    payload: opts.user,
    scope: opts.scope,
  });

  const { value, hit } = await withCache(key, async () => {
    const result = await structured({
      system: opts.system,
      user: opts.user,
      schema: opts.schema,
      schemaName: opts.schemaName,
      step: opts.step,
      signal: opts.signal,
    });
    return { data: result.value, usage: result.usage };
  });

  const payload = value as { data: T; usage: Usage };

  return {
    value: payload.data,
    trace: {
      step: opts.step,
      model: config.model,
      ms: Date.now() - started,
      usage: hit ? null : payload.usage,
      cacheHit: hit,
    },
  };
}

/* --------------------------------------------------------------- extract  */

export function extractResumeFacts(
  rawResumeText: string,
  scope: CacheScope,
  signal?: AbortSignal,
): Promise<Traced<ResumeFacts>> {
  return runStep<ResumeFacts>({
    step: "extract",
    promptVersion: PROMPTS.extract.version,
    system: PROMPTS.extract.text,
    user: `Extract structured facts from this resume.\n\n${block("resume", rawResumeText)}`,
    schema: ResumeFactsSchema,
    schemaName: "ResumeFacts",
    // A resume is the most sensitive thing in the system. Never shared.
    scope,
    signal,
  });
}

/* -------------------------------------------------------------------- jd  */

export function extractJobSpec(
  jdText: string,
  sourceNote: string,
  signal?: AbortSignal,
): Promise<Traced<JobSpec>> {
  return runStep<JobSpec>({
    step: "analyze",
    promptVersion: PROMPTS.jd.version,
    system: PROMPTS.jd.text,
    user: `Decompose this job description.\n\nSource: ${sourceNote}\n\n${block("job_description", jdText)}`,
    schema: JobSpecSchema,
    schemaName: "JobSpec",
    // A public job posting carries nothing private, and two candidates
    // applying to the same role should pay for one decomposition between them.
    scope: { kind: "global" },
    signal,
  });
}

/* ------------------------------------------------------------------ gaps  */

export function analyzeGaps(
  job: JobSpec,
  facts: ResumeFacts,
  scope: CacheScope,
  signal?: AbortSignal,
): Promise<Traced<GapAnalysis>> {
  return runStep<GapAnalysis>({
    step: "gaps",
    promptVersion: PROMPTS.gaps.version,
    system: PROMPTS.gaps.text,
    user: [
      "Compare this candidate against this role.",
      block("job_spec", JSON.stringify(job, null, 2)),
      block("resume_facts", JSON.stringify(facts, null, 2)),
    ].join("\n\n"),
    schema: GapAnalysisSchema,
    schemaName: "GapAnalysis",
    scope,
    signal,
  });
}

/* ---------------------------------------------------------------- tailor  */

export type TailorOutcome = {
  tailored: TailoredResume;
  truth: TruthReport;
  report: AtsReport;
  repairAttempted: boolean;
  /** Why the first draft was rejected. The repair doubles the cost of the most
   *  expensive step, so which rule keeps tripping is the difference between
   *  tuning the prompt and guessing at it. */
  firstDraftViolations: string[];
  traces: StepTrace[];
};

function buildBrief(job: JobSpec, facts: ResumeFacts, gaps: GapAnalysis): string {
  return [
    "Rewrite this candidate's resume for this role.",
    block("job_spec", JSON.stringify(job, null, 2)),
    block("resume_facts", JSON.stringify(facts, null, 2)),
    block("gap_analysis", JSON.stringify(gaps, null, 2)),
    "The gap analysis lists missing_keywords the candidate cannot truthfully claim. " +
      "Do not use them.\n" +
      "Surface every recoverable_keyword - those are already true and merely buried.",
  ].join("\n\n");
}

/**
 * Rewrite, verify, and repair exactly once.
 *
 * One repair, not a loop. A loop that keeps pushing until the guard passes is
 * an optimiser applying pressure toward whatever wording slips past it, which
 * is precisely the opposite of what the guard is for. If a second draft still
 * fails, the candidate is shown which lines cannot be traced rather than
 * handed a resume that reads well and cannot be defended in an interview.
 */
export async function tailorResume(args: {
  job: JobSpec;
  facts: ResumeFacts;
  gaps: GapAnalysis;
  rawResumeText: string;
  scope: CacheScope;
  signal?: AbortSignal;
  onEvent?: (event: { type: "draft" | "verified" | "repairing" }) => void;
}): Promise<TailorOutcome> {
  const { job, facts, gaps, rawResumeText, scope, signal } = args;
  const jdTerms = [
    ...job.requirements.map((r) => r.term),
    ...job.keywords.flatMap((k) => [k.term, ...k.variants]),
  ];
  const traces: StepTrace[] = [];
  const brief = buildBrief(job, facts, gaps);

  const first = await runStep<TailoredResume>({
    step: "tailor",
    promptVersion: PROMPTS.tailor.version,
    system: PROMPTS.tailor.text,
    user: brief,
    schema: TailoredResumeSchema,
    schemaName: "TailoredResume",
    scope,
    signal,
  });
  traces.push(first.trace);
  args.onEvent?.({ type: "draft" });

  let tailored = sanitize(first.value);
  let truth = await runGuard({ tailored, facts, rawResumeText, jdTerms, entailment: true });
  args.onEvent?.({ type: "verified" });

  let repairAttempted = false;
  let firstDraftViolations: string[] = [];

  if (!truth.passed) {
    repairAttempted = true;
    args.onEvent?.({ type: "repairing" });

    const errors = truth.violations.filter((v) => v.severity === "error");
    firstDraftViolations = errors.map((v) => `${v.code}: ${v.detail}`);
    const findings = errors
      .map((v) => `- [${v.code}] ${v.location}: ${v.detail}\n  Line: "${v.offending}"`)
      .join("\n");

    console.warn(
      `[tailor] guard rejected the first draft (${truth.error_count} errors): ` +
        firstDraftViolations.slice(0, 5).join("; "),
    );

    const repair = await runStep<TailoredResume>({
      step: "tailor",
      promptVersion: PROMPTS.tailor.version,
      system: PROMPTS.tailor.text,
      user: [
        brief,
        `Your previous draft failed verification against the uploaded resume:\n${findings}`,
        "Rewrite it. For each failing line, either restate it using only what its " +
          "sources actually say, or drop it. Do not try to justify a figure or a " +
          "technology that is not in the original resume.",
      ].join("\n\n"),
      schema: TailoredResumeSchema,
      schemaName: "TailoredResume",
      scope,
      signal,
    });
    traces.push(repair.trace);

    tailored = sanitize(repair.value);
    truth = await runGuard({ tailored, facts, rawResumeText, jdTerms, entailment: true });
  }

  // Computed, not asked of a model: free, instant, and the same answer every
  // time, which is what lets the editor re-score on every keystroke.
  const report = computeAtsReport(job, facts, tailored);

  return { tailored, truth, report, repairAttempted, firstDraftViolations, traces };
}
